src/daft-catalog/src/impls/memory.rs

```diff
@@ -0,0 +1,368 @@
+
+        self.tables
+            .read()
+            .unwrap()
+            .get(&namespace)
+            .and_then(|namespace_tables| {
+                namespace_tables
+                    .get(table_name)
+                    .map(|t| t.clone() as TableRef)
+            })
+            .ok_or_else(|| CatalogError::obj_not_found("table", ident))
+    }
+
+    fn has_namespace(&self, ident: &Identifier) -> CatalogResult<bool> {
+        if ident.has_qualifier() {
+            return Ok(false);
+        }
+
+        let namespace = ident.name();
+
+        Ok(self
+            .tables
+            .read()
+            .unwrap()
+            .contains_key(&Some(namespace.to_string())))
+    }
+
+    fn has_table(&self, ident: &Identifier) -> CatalogResult<bool> {
+        let Ok((namespace, table_name)) = Self::split_table_ident(ident) else {
+            return Ok(false);
+        };
+
+        Ok(self
+            .tables
+            .read()
+            .unwrap()
+            .get(&namespace)
+            .is_some_and(|namespace_tables| namespace_tables.contains_key(table_name)))
+    }
+
+    fn list_namespaces(&self, pattern: Option<&str>) -> CatalogResult<Vec<Identifier>> {
+        if pattern.is_some() {
+            return Err(CatalogError::unsupported(
+                "MemoryCatalog.list_namespaces does not support specifying a pattern.",
+            ));
+        }
+
+        Ok(self
+            .tables
+            .read()
+            .unwrap()
+            .keys()
+            .filter_map(|namespace| namespace.as_ref().map(Identifier::simple))
+            .collect())
+    }
+
+    fn list_tables(&self, pattern: Option<&str>) -> CatalogResult<Vec<Identifier>> {
+        let tables = self.tables.read().unwrap();
+        if let Some(pat) = pattern {
+            if let Some(namespace_tables) = tables.get(&Some(pat.to_string())) {
+                Ok(namespace_tables
+                    .keys()
+                    .map(|table_name| Identifier::new(vec![pat, table_name]))
+                    .collect())
+            } else {
+                Ok(vec![])
+            }
+        } else {
+            Ok(tables
+                .iter()
+                .flat_map(|(namespace, namespace_tables)| {
+                    namespace_tables
+                        .keys()
+                        .map(move |table_name| match namespace {
+                            Some(ns) => Identifier::new(vec![ns, table_name]),
+                            None => Identifier::simple(table_name),
+                        })
+                })
+                .collect())
+        }
+    }
+
+    #[cfg(feature = "python")]
+    fn to_py(&self, py: pyo3::Python<'_>) -> pyo3::PyResult<pyo3::PyObject> {
+        use pyo3::{intern, types::PyAnyMethods};
+
+        use crate::python::PyCatalog;
+
+        let pycatalog = PyCatalog(Arc::new(self.clone()));
+
+        Ok(py
+            .import(intern!(py, "daft.catalog.__rust"))?
```

Comment:
nit: super nit, but for some reason the `__rust` throws me off vs like `__internal` - something about mixing Py and `__rust` and encoding an implementation detail in the module naming itself. Would be different if `daft.daft` were `daft.rust` or `daft.__rust`.

Reply:
I'll rename to __internal